[OKTA-705854] Policies doc upate (#4714)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/okta-expression-language/index.md

This document details the features and syntax of the Okta Expression Language (EL). You can use this language throughout the Okta Admin Console and API for the Okta Classic Engine and Okta Identity Engine.

> **Note:** If you're using EL for the [authentication policies](/docs/guides/configure-signon-policy/main/) of the Identity Engine, the [Access Certification campaigns](https://help.okta.com/okta_help.htm?id=ext-el-eg) or Entitlement Management policies for Okta Identity Governance, use the features and syntax of the [Okta Expression Language in Okta Identity Engine](/docs/reference/okta-expression-language-in-identity-engine/).

Okta Expression Language is based on [SpEL](http://docs.spring.io/spring/docs/3.0.x/reference/expressions.html) and uses a subset of the functionalities offered by SpEL.

> Expressions allow you to reference, transform, and combine attributes before storing them on a user profile or passing them to an application for authentication or provisioning. For example, you might use a custom expression to create a username by stripping `@company.com` from an email address. Or, you might combine the `firstName` and `lastName` attributes into a single `displayName` attribute.

Reviewer: ```suggestion
Expressions allow you to reference, transform, and combine attributes before storing them on a user profile or passing them to an app for authentication or provisioning. For example, you might use a custom expression to create a username by stripping `@company.com` from an email address. Or, you might combine the `firstName` and `lastName` attributes into a single `displayName` attribute.
```

## Reference user attributes

When you create an Okta expression, you can reference any attribute that lives on an Okta user profile or application user profile.

### Okta user profile

Every user has an Okta user profile. The Okta user profile is the central source of truth for the core attributes of a user. To reference an Okta user profile attribute, specify `user.` and the attribute variable name. For a list of core user profile attributes, see [Default profile properties](/docs/reference/api/users/#default-profile-properties).

| Syntax            | Definitions                                                                   | examples                                                       |
| --------          | ----------                                                                    | ------------                                                   |
| `user.$attribute` | `user` reference to the Okta user<br>`$attribute` the attribute variable name | user.firstName<br>user.lastName<br>user.login<br>user.email |

> **Note:** You can also access the user ID for each user with the following expression: `user.getInternalProperty("id")`.

### Application user profile

> In addition to an Okta user profile, all users have a separate application user profile for each of their applications. Application user profiles store application-specific information about users, such as the application `userName` or user `role`.

Reviewer: ```suggestion
In addition to an Okta user profile, all users have a separate application user profile for each of their apps. Application user profiles store app-specific information about users, such as the app `userName` or user `role`.
```

Reviewer: Leaving this instance of "application" because it sounds like an "application user profile" is a specific thing?

Author: I checked the docs and "app user" and "app user profile" seem to be regularly used. I'll change it!

> To reference an profile attribute of an app user, specify the application variable and the attribute variable in the user profile of the application. In specifying the application, you can either name the specific application you're referencing or use an implicit reference to an in-context application.

Reviewer: ```suggestion
To reference an profile attribute of an app user, specify the app variable and the attribute variable in the user profile of the app. In specifying the app, you can either name the specific application you're referencing or use an implicit reference to an in-context app.
```

> > **Note:** The application reference is usually the `name` of the application, as distinct from the `label` (display name). See [Application properties](/docs/reference/api/apps/#application-properties). If your organization configures multiple instances of the same application, a randomly assigned suffix differentiates the names of the subsequent instances, for example: `zendesk_9ao1g13`. The name of any specific app instance in the Profile Editor appears in lighter text beneath the label of the app.

Reviewer: ```suggestion
> **Note:** The app reference is usually the `name` of the app, as distinct from the `label` (display name). See [Application properties](/docs/reference/api/apps/#application-properties). If your organization configures multiple instances of the same app, a randomly assigned suffix differentiates the names of the subsequent instances, for example: `zendesk_9ao1g13`. The name of any specific app instance in the Profile Editor appears in lighter text beneath the label of the app.
```

### IdP user profile

In addition to an Okta user profile, some users have separate IdP user profiles for their external Identity Provider. These IdP user profiles are used to store IdP-specific information about a user. You can use this data in an EL expression to transform an external user's username into the equivalent Okta username.

To reference an IdP user profile attribute, specify the IdP variable and the corresponding attribute variable for the IdP user profile of that Identity Provider. This profile is only available when specifying the username transform used to generate an Okta username for the IdP user.

| Syntax                 | Definitions                                                                                  | Examples          |
| ---------------------- | -------------------------------------------------------------------------------------------- | ------------      |
| `idpuser.$attribute`   | `idpuser` implicit reference to in-context IdP<br>`$attribute` the attribute variable name   | idpuser.firstName |

> > **Note:** In the Universal Directory, the base Okta user profile has about 30 attributes. You can add any number of custom attributes. All application user profiles have a username attribute and possibly others depending on the application. To find a full list of Okta user and app user attributes and their variable names, in the Admin Console go to **People** > **Profile Editor**. If you're not using Universal Directory, contact your support or professional services team.

Reviewer: ```suggestion
> **Note:** In the Universal Directory, the base Okta user profile has about 30 attributes. You can add any number of custom attributes. All application user profiles have a username attribute and possibly others depending on the app. To find a full list of Okta user and app user attributes and their variable names, in the Admin Console go to **People** > **Profile Editor**. If you're not using Universal Directory, contact your support or professional services team.
```

## Reference application and organization properties

> In addition to referencing user attributes, you can also reference application properties and the properties of your organization. To reference a particular attribute, specify the appropriate binding and the attribute variable name. The binding for an application is its name with `_app` appended. The app name can be found as described in the [profile attributes of an app user](#application-user-profile).

Reviewer: ```suggestion
In addition to referencing user attributes, you can also reference app properties and the properties of your org. To reference a particular attribute, specify the appropriate binding and the attribute variable name. The binding for an application is its name with `_app` appended. The app name can be found as described in the [profile attributes of an app user](#application-user-profile).
```

| IF            | !user.isContractor                             | fullTimeEmployees                              |
| IF            | user.salary > 1000000 AND !user.isContractor   | expensiveFullTimeEmployees                     |

> #### Null and blank attributes checking examples

Reviewer 2: ```suggestion
#### Check for null and blank attributes
```

Group rule conditions only allow `String`, `Arrays`, and `user` expressions.

> For example, given the user profile has a base string attribute called `email`, and assuming the user profile has a custom boolean attribute called `hasBadge` and a custom string attribute called `favoriteColor`, the following expressions are allowed in group rule conditions:

Reviewer: ```suggestion
For example, given the user profile has a base string attribute called `email`, and assuming the user profile has a custom Boolean attribute called `hasBadge` and a custom string attribute called `favoriteColor`, the following expressions are allowed in group rule conditions:
```
